File: 238-ProductOfArrayExceptSelf/238-ProductOfArrayExceptSelf.java
// Last updated: 8/13/2025, 8:20:48 PM
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] p = new int[n];
        p[0] = nums[0];
        for(int i=1;i<n;i++){
            p[i] = p[i-1] * nums[i];
        }
        int suffix = 1;
        for(int i=n-1;i>0;i--){
            p[i] = p[i-1] * suffix;
            suffix = suffix * nums[i];
        }
        p[0] = suffix;
        return p;
    }
}

File: 238-ProductOfArrayExceptSelf/238-ProductOfArrayExceptSelf.py
# Last updated: 8/13/2025, 8:20:42 PM
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)

        runningProduct = [None]*n
        backwardsProduct = [None]*n

        runningProduct[0] = nums[0]
        for i in range(1 , n) : 
            runningProduct[i] = runningProduct[i-1] * nums[i]
        
        backwardsProduct[n-1] = nums[n-1]
        for i in range(n-2 , -1 , -1) : 
            backwardsProduct[i] = backwardsProduct[i+1] * nums[i]
        
        multiplier = 1
        for i in range(n-1 , 0 , -1) : 
            temp = backwardsProduct[i]
            backwardsProduct[i] = multiplier * runningProduct[i-1]
            multiplier = temp
        
        backwardsProduct[0] = multiplier
        return backwardsProduct